### DISCUSSION TITLE
Language Idea: new override on methods

### DISCUSSION THREAD

**Discussion by hamarb123** [dotnet/csharplang#6697] [Language Ideas]
The language idea I want to propose is being able to put `new override` on methods.

Currently there is no good way to both override and shadow a method at the same time, which makes sense on the surface, but there are good reasons you may want to do both without any good way of doing them.

One scenario is if you have an external base class like so:
```csharp
public class SomeBaseClass
{
    public virtual int Get(A value) => ...;
    public int Get(B value) => ...;
}
```

Where `A` is implicitly convertible to `B`.
And where the implementation with `B` falls back to the implementation of `A` in a non-trivial way (meaning expensive).

This is fine, because we can do the following:
```csharp
A a = new();
B b = new();
SomeBaseClass instance = new();
instance.Get(a); //calls int Get(A)
instance.Get(b); //calls int Get(B)
```

If you want to provide a new implementations for both in our own derived class, it's relatively easy:
```csharp
public class SomeInheritedClass : SomeBaseClass
{
    public override int Get(A value) => ...;
    public new int Get(B value) => ...;
}
```

Except there is a big issue with the above code that is not at all obvious, no matter how hard you try, on an instance of `SomeInheritedClass` you cannot call `int Get(A)` without completely hiding the override with `B` from the compiler with something like `((SomeBaseClass)instance).Get(a)`.

i.e.
```csharp
A a = new();
B b = new();
SomeInheritedClass instance = new();
instance.Get(a); //calls int Get(B)
instance.Get(b); //calls int Get(B)
```

This is bad as callers may not realise they have to go out of their way to call an optimised overload since they provided exactly the type of the efficient overload. I only noticed this issue in my code when I was doing the final double check of the test code and was like "why is it using that overload?".

This is caused by method overload resolution: when selecting the most appropriate overload, it looks at `newslot` methods in the type and base types (ignoring ones with `PreserveBaseOverridesAttribute` for the purposes described here apparently) for which the parameters match or are implicitly convertible, and if multiple methods match, it prioritises the one in a more derived class, rather than prioritising ones with the correct arguments (which is obviously behaviour we cannot change now). (or something very similar to what I've just described)

The point is, there is no good way to make it so your callers to easily use the correct overload. You could theoretically work around it by adding an extra in-between class, but this is obviously unideal since the code is unclear, and bad:
```csharp
public abstract class SomeInheritedClassHelper : SomeBaseClass
{
    public override int Get(A value) => GetHelper(value);
    protected abstract int GetHelper(A value);
}
public class SomeInheritedClass : SomeInheritedClassHelper
{
    protected override int GetHelper(A value) => ...;
    public new int Get(A value) => GetHelper(value);
    public new int Get(B value) => ...;
}
```

The above solution also suggests what a good solution should look like. The `new override` modifier combination would make the method overloading treat it as if it is a `newslot` method (even ignoring `PreserveBaseOverridesAttribute` for these purposes), whilst still overriding the base implementation.

There are 2 logical ways of implementing it that I can think of:

1. A attribute generated by the C# compiler e.g. `System.Runtime.InteropServices.NewOverrideAttribute` (or something similar) - since this is a compiler-only feature, this should be fine.
2. It could emit IL that does this, it is already effectively supported:
```csharp
public new override void M()
{
    base.M();
}
```
Turns into:
```il
.method public hidebysig newslot virtual 
    instance void M () cil managed 
{
    .maxstack 8

    //this is the one with actual implementation
    ldarg.0
    call instance void C1::M()
    ret
}

.method public final hidebysig virtual
    instance void '$NewOverride$M' () cil managed 
{
    .maxstack 8
    .override method instance void class C1::M()

    ldarg.0
    call instance void C2::M()
    ret
}
```

The second approach has the advantage of also working on older compilers consuming these, and with other languages consuming these ([here](https://sharplab.io/#v2:EYLgtghglgdgNAGxFBAfAAgNgAQEECwAUAN5EC+RR6A7NgMICMR2L26AHEaYa9gPR9sAWQCmAFwAWAewAmAZ2at0ATmwwRAdzkIpY7ADcoAJzEBXCAmyLeLWHLEQYAYxEGpUGcOwAKAJRsAZiseVm4bVgFhcWlPYBEAcztsCD0AJQA1XGwABgAPACZszGprG0i6WVc5KAAvVwYfPIZfUt50AFZsTkJW1gBJABkAfWzR7JBsI3FSsn5BERhPKQAzbDBo2XoQECFySkIAOicECDk5bAAHU2AEKCdk0zEpZJhq7DjlqSnlqBEEGVgUD0dHypREuTEC3k9CYJCoqnUWh0ekMJnMllKdgczlc+ncniEPn86CCXFKHS6+14/wgRniB2ypScFksWMcLjcHhh2yEflKUzEex6h3Wkk2Vxud2wPxgFmwEg8ImAAE9qvEDMYzBZMa9sRy8VyAOQAEgAcpoAPL6ERGIyK41CQ1EwLBMJKTrdXgHKTW22KtYbTxsnGczzHU7nRg8vnC6kyWn0xkhFjMhCs3Xs3H4+j5aMtZOTaaECjFoA===)'s an example, C# decompilation is obviously wrong, but shows what it's doing).

If approved, this should also be compatible with other modifiers like `abstract`.

Thanks for reading through all of this :)

**Comment by iam3yal** [comment]
@hamarb123 I'm not sure why you would want to shadow the method in this case but you can always specialize and do something like this:
```csharp
public override int Get(A value) => ...;
public new int Get(B value) => ...;

public int Get<T>(T t) where T : B
{
    if (t is A a)
    {
        return ((SomeBaseClass)this).Get(a);
    }
    
    return Get(t);
}
```
You can also implement it directly on `Get(B value)` as opposed to defining a new generic method because in this case `A` derives from `B` like this:
```csharp
public new int Get(B value)
{
    if (value is A a)
    {
        return ((SomeBaseClass)this).Get(a);
    }
    
    return ...;
}
```

**Reply by hamarb123** [reply]
> I'm not sure why you would want to shadow the method in this case but you can always specialize and do something like this:

It's not about shadowing itself, it's about changing the overload resolution, which is currently done based on which methods are `newslot`. The first implementation I suggested doesn't even make the method a `newslot` method, because it's not really the point.

> Where `A` is implicitly convertible to `B`.
And where the implementation with `B` falls back to the implementation of `A` in a non-trivial way (meaning expensive).

(The above quote is from my initial post, pointing out when this is most useful)
If it were just inheritance, that would work, but I'm more referring to some sort of conversion that can't easily be undone, e.g. `byte[]` -> `ReadOnlySpan<byte>`, `string` -> `ReadOnlySpan<char>`, these conversions can absolutely not be undone. This can be the case often if you're deferring to an older or less-spanified library for example, if you're lucky they may provide a pointer overload which means you *only* have to pin a massive chunk of memory rather than copy it (which is obviously still not ideal). Luckily in my specific case, I found a way to work around it (by implementing the relevant API with span without having to pin it, by using `unsafe` code), but this may not always be possible e.g. when you need to store it to a field you obviously have to copy the span, but you may still want to provide a catch-all overload, or for compatibility reasons. I've only provided examples with spans here, but the issue is much more general, since `implicit` conversions are common even in cases with some information loss if they are to something more generic (e.g., array -> span).

In fact I'd argue that in the general case, `implicit` conversions mostly lose information (or keep it the same) in one way or another, e.g. `byte` -> `ushort` actually loses you 8 bits of information compared to the range of possible values you knew previously, and you'd have to add another check for that range if you wanted to special-case it - this type of special-case checking is not always possible as with the array->span case since it's not reversible.
